File: challenges/groceries/bot/planner.py
"""
Per-bot state machine.

States:
  COLLECTING   — navigate to assigned item and pick it up
  DELIVERING   — navigate to drop-off zone and deliver
  PRE_FETCHING — pick up preview-order items when we have spare slots

Inventory fill rule:
  Don't deliver after 1 item — fill to min(3, remaining_needed) first,
  unless end-game or already at drop-off.
"""
from __future__ import annotations
from typing import Optional
from enum import Enum, auto

from state import Bot, Item, GameState, Order
from pathfinding import next_step_toward, precompute_drop_distances


class BotState(Enum):
    COLLECTING = auto()
    DELIVERING = auto()
    PRE_FETCHING = auto()
    IDLE = auto()


def nearest_dropoff(
    pos: tuple[int, int],
    drop_off_zones: list[tuple[int, int]],
    bot_counts: dict[tuple[int, int], int],
) -> tuple[int, int]:
    """
    Returns the least-congested nearest drop-off zone.
    Congestion penalty: 2 per bot within radius 3 of the zone.
    """
    def cost(zone):
        from pathfinding import bfs_distance
        # We only have position here, use Manhattan for tie-breaking
        manhattan = abs(pos[0] - zone[0]) + abs(pos[1] - zone[1])
        congestion = bot_counts.get(zone, 0) * 2
        return manhattan + congestion

    return min(drop_off_zones, key=cost)


def plan_bot_action(
    bot: Bot,
    state: GameState,
    assigned_item: Optional[Item],
    drop_dist: dict[tuple[int, int], int],
    walls: frozenset[tuple[int, int]],
    blocked: frozenset[tuple[int, int]],
    dropoff_congestion: dict[tuple[int, int], int],
) -> dict:
    """
    Returns an action dict for this bot:
      {"action": "move_up"} or {"action": "pick_up", "item_id": "item_0"} etc.
    """
    pos = bot.position
    inventory = bot.inventory
    active = state.active_order
    preview = state.preview_order
    drop_zones = state.drop_off_zones
    w, h = state.grid.width, state.grid.height

    # ── Determine what's useful to deliver ──────────────────────────────────
    active_needed = active.still_needed if active else []
    useful_in_inv = [i for i in inventory if i in active_needed]

    # ── Check if we're standing on a drop-off zone ──────────────────────────
    on_dropoff = pos in drop_zones

    # ── END-GAME: deliver immediately if holding anything useful ────────────
    if state.end_game and useful_in_inv:
        if on_dropoff:
            return {"action": "drop_off"}
        target = nearest_dropoff(pos, drop_zones, dropoff_congestion)
        step = next_step_toward(pos, target, walls, w, h, blocked)
        return {"action": step or "wait"}

    # ── DELIVERING: if holding useful items and should go deliver ────────────
    if useful_in_inv:
        # Inventory fill check: should we pick up more first?
        truly_needed = state.truly_needed(active)
        should_fill_more = (
            len(inventory) < 3
            and len(truly_needed) > 0
            and assigned_item is not None
            and not state.end_game
        )

        if not should_fill_more:
            # Go deliver
            if on_dropoff:
                return {"action": "drop_off"}
            target = nearest_dropoff(pos, drop_zones, dropoff_congestion)
            step = next_step_toward(pos, target, walls, w, h, blocked)
            return {"action": step or "wait"}

    # ── COLLECTING: go pick up assigned item ────────────────────────────────
    if assigned_item is not None and len(inventory) < 3:
        item_pos = assigned_item.position
        dist = abs(pos[0] - item_pos[0]) + abs(pos[1] - item_pos[1])

        if dist == 1:
            # Adjacent — pick up
            return {"action": "pick_up", "item_id": assigned_item.id}
        else:
            step = next_step_toward(pos, item_pos, walls, w, h, blocked)
            return {"action": step or "wait"}

    # ── PRE_FETCHING: pick up preview items if we have slots ────────────────
    if preview and len(inventory) < 3:
        preview_needed_types = preview.still_needed
        preview_items = state.available_items_for(preview_needed_types)

        # Only pre-fetch if the active order is nearly done
        truly_needed = state.truly_needed(active)
        if len(truly_needed) <= 1 and preview_items:
            # Find nearest preview item
            best = min(
                preview_items,
                key=lambda i: abs(pos[0] - i.position[0]) + abs(pos[1] - i.position[1]),
            )
            dist = abs(pos[0] - best.position[0]) + abs(pos[1] - best.position[1])
            if dist == 1:
                return {"action": "pick_up", "item_id": best.id}
            step = next_step_toward(pos, best.position, walls, w, h, blocked)
            if step:
                return {"action": step}

    # ── IDLE: move toward drop-off if holding anything, else wait ────────────
    if inventory:
        if on_dropoff:
            return {"action": "drop_off"}
        target = nearest_dropoff(pos, drop_zones, dropoff_congestion)
        step = next_step_toward(pos, target, walls, w, h, blocked)
        return {"action": step or "wait"}

    return {"action": "wait"}

File: challenges/groceries/bot/planner.ts
import type { Bot, Item, GameState, Position, BotAction, DistanceMap } from "./types.ts";
import { posKey, positionsInclude, manhattanDistance } from "./utils.ts";
import { nextStepToward } from "./pathfinding.ts";
import { stillNeeded, getActiveOrder, getPreviewOrder, trulyNeeded, availableItemsFor, isEndGame } from "./state.ts";

export type BotStateName = "collecting" | "delivering" | "pre_fetching" | "idle";

export function nearestDropoff(
  pos: Position,
  dropOffZones: Position[],
  botCounts: Map<string, number>,
): Position {
  let bestZone = dropOffZones[0];
  let bestCost = Infinity;

  for (const zone of dropOffZones) {
    const manhattan = manhattanDistance(pos, zone);
    const congestion = (botCounts.get(posKey(zone)) ?? 0) * 2;
    const cost = manhattan + congestion;
    if (cost < bestCost) {
      bestCost = cost;
      bestZone = zone;
    }
  }

  return bestZone;
}

export function planBotAction(
  bot: Bot,
  state: GameState,
  assignedItem: Item | null,
  dropDist: DistanceMap,
  walls: Set<string>,
  blocked: Set<string>,
  dropoffCongestion: Map<string, number>,
): BotAction {
  const pos = bot.position;
  const inventory = bot.inventory;
  const active = getActiveOrder(state);
  const preview = getPreviewOrder(state);
  const dropZones = state.dropOffZones;
  const { width, height } = state.grid;

  // Determine what's useful to deliver
  const activeNeeded = active ? stillNeeded(active) : [];
  const usefulInInv = inventory.filter((item) => activeNeeded.includes(item));

  // Check if we're standing on a drop-off zone
  const onDropoff = positionsInclude(dropZones, pos);

  // END-GAME: deliver immediately if holding anything useful
  if (isEndGame(state) && usefulInInv.length > 0) {
    if (onDropoff) return { action: "drop_off" };
    const target = nearestDropoff(pos, dropZones, dropoffCongestion);
    const step = nextStepToward(pos, target, walls, width, height, blocked);
    return { action: step ?? "wait" };
  }

  // DELIVERING: if holding useful items and should go deliver
  if (usefulInInv.length > 0) {
    const trulyNeededItems = trulyNeeded(state, active);
    const shouldFillMore =
      inventory.length < 3 &&
      trulyNeededItems.length > 0 &&
      assignedItem !== null &&
      !isEndGame(state);

    if (!shouldFillMore) {
      if (onDropoff) return { action: "drop_off" };
      const target = nearestDropoff(pos, dropZones, dropoffCongestion);
      const step = nextStepToward(pos, target, walls, width, height, blocked);
      return { action: step ?? "wait" };
    }
  }

  // COLLECTING: go pick up assigned item
  if (assignedItem !== null && inventory.length < 3) {
    const itemPos = assignedItem.position;
    const dist = manhattanDistance(pos, itemPos);

    if (dist === 1) {
      return { action: "pick_up", item_id: assignedItem.id };
    }
    const step = nextStepToward(pos, itemPos, walls, width, height, blocked);
    return { action: step ?? "wait" };
  }

  // PRE_FETCHING: pick up preview items if we have slots
  if (preview && inventory.length < 3) {
    const previewNeededTypes = stillNeeded(preview);
    const previewItems = availableItemsFor(state, previewNeededTypes);

    const trulyNeededItems = trulyNeeded(state, active);
    if (trulyNeededItems.length <= 1 && previewItems.length > 0) {
      // Find nearest preview item
      const best = previewItems.reduce((closest, item) =>
        manhattanDistance(pos, item.position) < manhattanDistance(pos, closest.position)
          ? item
          : closest,
      );
      const dist = manhattanDistance(pos, best.position);
      if (dist === 1) {
        return { action: "pick_up", item_id: best.id };
      }
      const step = nextStepToward(pos, best.position, walls, width, height, blocked);
      if (step) return { action: step };
    }
  }

  // IDLE: move toward drop-off if holding anything, else wait
  if (inventory.length > 0) {
    if (onDropoff) return { action: "drop_off" };
    const target = nearestDropoff(pos, dropZones, dropoffCongestion);
    const step = nextStepToward(pos, target, walls, width, height, blocked);
    return { action: step ?? "wait" };
  }

  return { action: "wait" };
}
